pydantic_ai_slim/pydantic_ai/ag_ui.py

```diff
@@ -0,0 +1,871 @@
+        AssistantMessage,
+        BaseEvent,
+        DeveloperMessage,
+        EventType,
+        Message,
+        RunAgentInput,
+        RunErrorEvent,
+        RunFinishedEvent,
+        RunStartedEvent,
+        State,
+        SystemMessage,
+        TextMessageContentEvent,
+        TextMessageEndEvent,
+        TextMessageStartEvent,
+        ThinkingTextMessageContentEvent,
+        ThinkingTextMessageEndEvent,
+        ThinkingTextMessageStartEvent,
+        ToolCallArgsEvent,
+        ToolCallEndEvent,
+        ToolCallResultEvent,
+        ToolCallStartEvent,
+        ToolMessage,
+        UserMessage,
+    )
+    from ag_ui.encoder import EventEncoder
+except ImportError as e:  # pragma: no cover
+    raise ImportError(
+        'Please install the `ag-ui-protocol` package to use `Agent.to_ag_ui()` method, '
+        'you can use the `ag-ui` optional group — `pip install "pydantic-ai-slim[ag-ui]"`'
+    ) from e
+
+try:
+    from starlette.applications import Starlette
+    from starlette.middleware import Middleware
+    from starlette.requests import Request
+    from starlette.responses import Response, StreamingResponse
+    from starlette.routing import BaseRoute
+    from starlette.types import ExceptionHandler, Lifespan
+except ImportError as e:  # pragma: no cover
+    raise ImportError(
+        'Please install the `starlette` package to use `Agent.to_ag_ui()` method, '
+        'you can use the `ag-ui` optional group — `pip install "pydantic-ai-slim[ag-ui]"`'
+    ) from e
+
+from pydantic import BaseModel, ValidationError
+
+from ._agent_graph import CallToolsNode, ModelRequestNode
+from .agent import Agent, RunOutputDataT
+from .messages import (
+    AgentStreamEvent,
+    FinalResultEvent,
+    FunctionToolResultEvent,
+    ModelMessage,
+    ModelRequest,
+    ModelResponse,
+    ModelResponsePart,
+    PartDeltaEvent,
+    PartStartEvent,
+    SystemPromptPart,
+    TextPart,
+    TextPartDelta,
+    ThinkingPart,
+    ThinkingPartDelta,
+    ToolCallPart,
+    ToolCallPartDelta,
+    ToolReturnPart,
+    UserPromptPart,
+)
+from .models import KnownModelName, Model
+from .output import DeferredToolCalls, OutputDataT, OutputSpec
+from .result import AgentStream
+from .settings import ModelSettings
+from .tools import AgentDepsT, ToolDefinition
+from .toolsets import AbstractToolset
+from .toolsets.deferred import DeferredToolset
+from .usage import Usage, UsageLimits
+
+if TYPE_CHECKING:
+    from collections.abc import AsyncGenerator
+
+    from ag_ui.encoder import EventEncoder
+
+    from pydantic_graph.nodes import End
+
+    from ._agent_graph import AgentNode
+    from .agent import AgentRun
+    from .result import FinalResult
+
+# Variables.
+_LOGGER: logging.Logger = logging.getLogger(__name__)
+
+# Constants.
+SSE_CONTENT_TYPE: Final[str] = 'text/event-stream'
+"""Content type header value for Server-Sent Events (SSE)."""
+
+
+class FastAGUI(Generic[AgentDepsT, OutputDataT], Starlette):
```

Comment:
I like the riff on FastA2A, but I think this is a bit too suggestive of it being Pydantic AI-agnostic (as FastA2A is!). Maybe just `AGUIApp`? Or `App` as we're already inside `ag_ui`?

Reply:
Naming things is hard 😄